### ISSUE TITLE
Current `rack` dependency version causes conflict with modern `rackup` versions

### ISSUE THREAD

**Issue by 64kramsystem** [sj26/mailcatcher#562]
The `rack` dependency version, `~>2.2`, causes conflicts with modern `rackup` versions (e.g. `2.1.0`); it should be verified and upgraded to a modern version (e.g. `~>3.1`).

**Comment by sj26** [comment]
How does it cause conflicts?

**Comment by 64kramsystem** [comment]
Try installing `rack` version ~>2.2, then the latest `rackup` version; you'll get a conflict on the `rackup` binary.

**Comment by sj26** [comment]
Ah yes, I can reproduce:

```sh
$ gem install mailcatcher
...
Successfully installed rack-2.2.9
...

$ gem install rackup
Fetching rack-3.1.7.gem
Fetching rackup-2.1.0.gem
Successfully installed rack-3.1.7
rackup's executable "rackup" conflicts with rack
Overwrite the executable? [yN]
ERROR:  Error installing rackup:
	"rackup" from rackup conflicts with installed executable from rack
```

Sounds like a bump is in order. It might be possible now, given the dependencies on thin and websocket-driver have been unwound a little.

**Comment by aried3r** [comment]
Do you think this would potentially solve https://github.com/sj26/mailcatcher/issues/560?

**Comment by sj26** [comment]
No. #560 is an error in how homebrew has packaged mailcatcher.

Workaround for that is don’t install mailcatcher with homebrew. Just:

```
gem install mailcatcher
```

**Comment by dsandstrom** [comment]
> No. #560 is an error in how homebrew has packaged mailcatcher.
> 
> Workaround for that is don’t install mailcatcher with homebrew. Just:
> 
> ```
> gem install mailcatcher
> ```

Results in 

```
rack's executable "rackup" conflicts with rackup
Overwrite the executable? [yN]
ERROR:  Error installing mailcatcher:
	"rackup" from rack conflicts with installed executable from rackup
```

I don't want to overwrite the executable so can't install

**Comment by sj26** [comment]
Just press “n” for now

**Comment by dsandstrom** [comment]
> Just press “n” for now

Right, that results in not installing the gem

**Comment by 64kramsystem** [comment]
> > No. #560 is an error in how homebrew has packaged mailcatcher.
> > Workaround for that is don’t install mailcatcher with homebrew. Just:
> > ```
> > gem install mailcatcher
> > ```
> 
> Results in
> 
> ```
> rack's executable "rackup" conflicts with rackup
> Overwrite the executable? [yN]
> ERROR:  Error installing mailcatcher:
> 	"rackup" from rack conflicts with installed executable from rackup
> ```
> 
> I don't want to overwrite the executable so can't install

If you can, uninstall all the versions of `rackup`, and install version 1.0; it won't conflict.

**Comment by sj26** [comment]
> > Just press “n” for now
> 
> Right, that results in not installing the gem

No, it should install the gem, it just won’t link that particular executable (which isn’t required, at least for mailcatcher)

**Comment by dsandstrom** [comment]
> If you can, uninstall all the versions of `rackup`, and install version 1.0; it won't conflict.

Trying to get mailcatcher to work with rack 3, rackup 2.1

**Comment by dsandstrom** [comment]
> > > Just press “n” for now
> > 
> > 
> > Right, that results in not installing the gem
> 
> No, it should install the gem, it just won’t link that particular executable (which isn’t required, at least for mailcatcher)

I must be missing something

**Comment by dsandstrom** [comment]
I started to look into what it would take to update this gem so it works with rack 3. Unfortunately, I couldn't even get gems to install.

The limiting factor starts with sprockets-sass. Sprockets v4 supports rack 3, but sprockets-sass hasn't been updated to support that version.

https://github.com/petebrowne/sprockets-sass/blob/master/sprockets-sass.gemspec#L39

**Comment by jameshibbard** [comment]
I ran into the same issue. The install failed because mailcatcher tried to install its own rackup executable, which conflicted with the existing one on my system. Selecting "n" aborted the install.

```shell
gem install mailcatcher
...
rack's executable "rackup" conflicts with rackup
Overwrite the executable? [yN]  n
ERROR:  Error installing mailcatcher:
	"rackup" from rack conflicts with installed executable from rackup
```

Rackup is pulled in by Rails itself, so uninstalling it would likely break something:

```shell
gem uninstall rackup        

You have requested to uninstall the gem:
	rackup-2.2.1

railties-8.1.0 depends on rackup (>= 1.0.0)
```

Instead of fighting the conflict, I switched to [Mailpit](https://github.com/axllent/mailpit), which ships as a standalone binary. You just drop the executable into your PATH and run it.  It caught mails immediately and worked exactly as expected.

I'm happy to provide any debugging info if it'll help.

**Comment by spaquet** [comment]
@jameshibbard if you are interested I just suggested a massive PR with a bunch of upgrade to operate more recent Rack, Sinatra and Thin
https://github.com/sj26/mailcatcher/pull/573

My fork is here if interested to pull from it while waiting for the maintainer to upgrade https://github.com/spaquet/mailcatcher
I also have docker (small footprint) that could be interesting for you: https://github.com/spaquet/docker-alpine-mailcatcher based on my fork.

**Comment by jameshibbard** [comment]
Thanks, I'll take a look when I have a minute.

**Comment by spaquet** [comment]
I also completely redesigned the UI and upgraded the javascript libraries at the same time. there were a lot of "old" packages still in use.

**Comment by jasonperrone** [comment]
@jameshibbard Mailpit is great, thanks for that recommendation.

**Comment by aarona** [comment]
Having the same issue running Ruby 4.0.1 and Rack (with `rackup -v`) 3.2.5. Mailpit seems to work fine for me. I love mailcatcher though and would rather support a rubyist. Been using it for at least a decade now.

@spaquet you mentioned a PR but looks like you closed it due to some corner case issues you tested and you'd resubmit it later? Any status on that? Would love to shoot you and @sj26 some $ as a thank you for your work if you plan on moving forward with those updates. Would love to see a revamped version with a dark mode possibly. I've also noticed that I often have to refresh the page it runs in after sending emails. Usually only autoloads new emails once and stops after I review them. I should post an issue for that but it doesn't bother me that much.

**Comment by spaquet** [comment]
@aarona the fork went too far to be merged but if you are interested https://github.com/spaquet/mailcatcher
